bluebinder.c

```diff
@@ -368,38 +369,49 @@ bluebinder_callbacks_transact(
     if (!g_strcmp0(iface, BINDER_BLUETOOTH_SERVICE_IFACE_CALLBACKS)) {
         if (code == 1) { /* initializationComplete */
             int result = 0;
+            GBinderLocalReply* reply = NULL;
 
             gbinder_remote_request_read_int32(req, &result);
 
-            if (result != 0) {
-                fprintf(stderr, "Bluetooth binder service failed\n");
-                g_main_loop_quit(proxy->loop);
+            do{
+
                 *status = GBINDER_STATUS_FAILED;
-                return NULL;
-            }
+                proxy->is_bt_failed_on_init = TRUE;
 
-            fprintf(stderr, "Binder interface initialized, opening virtual device\n");
+                if (result != 0) {
+                    fprintf(stderr, "Bluetooth binder service failed\n");
+                    /* we need to tell BT service that we properly handled Status::INITIALIZATION_ERROR */
+                    *status = GBINDER_STATUS_OK;
+                    reply = gbinder_local_reply_append_int32(gbinder_local_object_new_reply(obj), 0);
+                    break;
+                }
 
-            proxy->host_fd = open_vhci(HCI_PRIMARY);
-            if (proxy->host_fd < 0) {
-                fprintf(stderr, "Unable to open virtual device\n");
-                g_main_loop_quit(proxy->loop);
-                *status = GBINDER_STATUS_FAILED;
-                return NULL;
-            }
+                fprintf(stderr, "Binder interface initialized, opening virtual device\n");
 
-            if (!setup_watch(proxy)) {
-                fprintf(stderr, "Unable to setup watch\n");
-                g_main_loop_quit(proxy->loop);
-                *status = GBINDER_STATUS_FAILED;
-                return NULL;
-            }
+                proxy->host_fd = open_vhci(HCI_PRIMARY);
+                if (proxy->host_fd < 0) {
+                    fprintf(stderr, "Unable to open virtual device\n");
+                    break;
+                }
 
-            sd_notify(0, "READY=1");
+                if (!setup_watch(proxy)) {
+                    fprintf(stderr, "Unable to setup watch\n");
+                    break;
+                }
 
-            *status = GBINDER_STATUS_OK;
+                sd_notify(0, "READY=1");
+
+                *status = GBINDER_STATUS_OK;
+                proxy->is_bt_failed_on_init = FALSE;
+                reply = gbinder_local_reply_append_int32(gbinder_local_object_new_reply(obj), 0);
+
+            } while (0);
+
+            if(proxy->is_bt_failed_on_init && proxy->loop)
+                g_main_loop_quit(proxy->loop);
+
+            return reply;
```

Comment:
Something like this
```c
return (*status == GBINDER_STATUS_OK) ?
    gbinder_local_reply_append_int32(gbinder_local_object_new_reply(obj), 0) :
    NULL;
```
would reduce the number of duplicate lines.

Reply:
agreed